Defer subclass methods if superclass has not been analyzed

ilevkivskyi opened the pull request:
Fixes #5560 
Fixes #5548 

Half of the PR is updating various function signatures to also accept `Decorator`. I still prohibit `Decorator` as a target in fine-grained mode, but I think there should be no harm to defer it in normal mode.

ilevkivskyi commented on mypy/checker.py:
```diff
@@ -338,20 +336,23 @@ def check_top_level(self, node: MypyFile) -> None:
+            type_name = self.errors.type_name[-1]
+        else:
+            type_name = None
+        # Shouldn't we freeze the entire scope?
```
I am not sure what was the intent. This comment is two years old, maybe just remove it?

JukkaL replied:
Maybe reword it like this: "We don't freeze the entire scope since only top-level functions and methods can be deferred. Only module/class level scope information is needed. Module-level scope information is preserved in the TypeChecker instance." (I *think* that this is correct.)

ilevkivskyi commented on mypy/server/update.py:
```diff
@@ -1058,7 +1058,7 @@ def is_verbose(manager: BuildManager) -> bool:
 
 def target_from_node(module: str,
-                     node: Union[FuncDef, MypyFile, OverloadedFuncDef, LambdaExpr]
+                     node: Union[FuncDef, MypyFile, OverloadedFuncDef, LambdaExpr, Decorator]
```
I can remove `Decorator` from all these unions, but this will require a couple of asserts.

JukkaL replied:
I think that it's better to have the caller ensure that the function is called with the correct types. The annotation as it is now is arguably misleading.

JukkaL replied:
Ah, good point! In that case maybe `LambdaExpr` should also be removed from the annotation?

ilevkivskyi replied:
Yes, I will remove both.

ilevkivskyi commented:
@JukkaL Thanks for review! I addressed the comments. I decided to have a separate `FineDeferredNode` named tuple (since named tuples can't be generic). This way we can avoid all asserts and extra items in unions in arguments.

I also added a bunch of tests combining overloads, decorators, variables, double deferrals, and super calls (also with various combinations of these).

JukkaL commented on mypy/checker.py:
```diff
@@ -1253,15 +1276,26 @@ def expand_typevars(self, defn: FuncItem,
+    def check_method_override(self, defn: Union[FuncDef, OverloadedFuncDef, Decorator]) -> None:
+        """Check if function definition is compatible with base classes.
+
+        This may defer the method if a signature is not available in at least on base class.
```
Typo: "on" -> "one"

JukkaL commented on test-data/unit/check-classes.test:
```diff
@@ -5117,3 +5117,295 @@ class C:
+def deco(f: Callable[..., T]) -> Callable[..., Tuple[T, int]]: ...
+[out]
+
+[case testCyclicDecoratorSuperDeferred]
```
Did you verify that these test cases trigger deferral? Using `from ... import` can be used to make it more explicit in which order modules are processed by mypy.

ilevkivskyi replied:
I actually checked only first two tests I added, I will also double-check rest just in case.

ilevkivskyi replied:
They all get deferred expected number of times (1, 2, or 3), I switched to `from ... import` in half of the tests anyway.

JukkaL commented on mypy/checker.py:
```diff
@@ -72,19 +73,31 @@
 DEFAULT_LAST_PASS = 1  # type: Final  # Pass numbers start at 0
 
+DeferredNodeType = Union[FuncDef, LambdaExpr, OverloadedFuncDef, Decorator]
+FineDeferredNodeType = Union[FuncDef, MypyFile, OverloadedFuncDef]
```
Style nit: I'd prefer to call this `FineGrainedDeferredNodeType` or something. 

We don't use "fine" as a shorted version of "fine-grained" elsewhere and it can be confusing.

JukkaL commented on mypy/checker.py:
```diff
@@ -72,19 +73,31 @@
 
+# Same as above, but for fine-grained mode targets. Only top-level functions/methods
+# and module top levels are allowed as such.
+FineDeferredNode = NamedTuple(
```
Similar to above, I'd prefer to name this `FineGrainedDeferredNode`.
